gcc/rust/expand/rust-macro-builtins.cc

```diff
@@ -63,4 +150,67 @@ MacroBuiltin::column (Location invoc_locus, AST::MacroInvocData &invoc)
 
   return AST::ASTFragment ({column_no});
 }
+
+/* Expand builtin macro include_bytes!("filename"), which includes the contents
+   of the given file as reference to a byte array. Yields an expression of type
+   &'static [u8; N].  */
+
+AST::ASTFragment
+MacroBuiltin::include_bytes (Location invoc_locus, AST::MacroInvocData &invoc)
+{
+  /* Get target filename from the macro invocation, which is treated as a path
+     relative to the include!-ing file (currently being compiled).  */
+  std::string include_path
+    = parse_single_string_literal (invoc.get_delim_tok_tree ())->as_string ();
+  std::string target_filename
+    = source_relative_path (include_path, invoc_locus);
+
+  std::vector<uint8_t> bytes = load_file_bytes (target_filename.c_str ());
+
+  /* Is there a more efficient way to do this?  */
+  std::vector<std::unique_ptr<AST::Expr>> elts;
+  for (uint8_t b : bytes)
+    {
+      elts.emplace_back (
+	new AST::LiteralExpr (std::string (1, (char) b), AST::Literal::BYTE,
+			      PrimitiveCoreType::CORETYPE_U8,
+			      {} /* outer_attrs */, invoc_locus));
+    }
+
+  auto elems = std::unique_ptr<AST::ArrayElems> (
+    new AST::ArrayElemsValues (std::move (elts), invoc_locus));
+
+  auto array = std::unique_ptr<AST::Expr> (
+    new AST::ArrayExpr (std::move (elems), {}, {}, invoc_locus));
+
+  auto borrow = std::unique_ptr<AST::Expr> (
+    new AST::BorrowExpr (std::move (array), false, false, {}, invoc_locus));
+
+  auto node = AST::SingleASTNode (std::move (borrow));
+  return AST::ASTFragment ({node});
+}
+
+/* Expand builtin macro include_str!("filename"), which includes the contents
+   of the given file as a string. The file must be UTF-8 encoded. Yields an
+   expression of type &'static str.  */
+
+AST::ASTFragment
+MacroBuiltin::include_str (Location invoc_locus, AST::MacroInvocData &invoc)
+{
+  /* Get target filename from the macro invocation, which is treated as a path
+     relative to the include!-ing file (currently being compiled).  */
+  std::string include_path
+    = parse_single_string_literal (invoc.get_delim_tok_tree ())->as_string ();
+  std::string target_filename
+    = source_relative_path (include_path, invoc_locus);
+
+  std::vector<uint8_t> bytes = load_file_bytes (target_filename.c_str ());
+
+  /* TODO: Enforce that the file contents are valid UTF-8?  */
```

Comment:
Yes! I think this is fine for now, but we should definitely open up an issue (and maybe turn this into a FIXME instead). rustc errors out if the included string is not valid UTF-8:

```rust
arthur@platypus ~/G/gccrs (laxer-parser-on-stmt-fragment) [1]> wget https://www.cl.cam.ac.uk/~mgk25/ucs/examples/UTF-8-test.txt -O incorrect.utf # snip
arthur@platypus ~/G/gccrs (laxer-parser-on-stmt-fragment)> cat test.rs
fn main() {
    println!("{}", include_str!("incorrect.utf"));
}
arthur@platypus ~/G/gccrs (laxer-parser-on-stmt-fragment)> rustc test.rs
error: incorrect.utf wasn't a utf-8 file
 --> test.rs:2:20
  |
2 |     println!("{}", include_str!("incorrect.utf"));
  |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `include_str` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error
```

Reply:
OK. Marked as FIXME, will open up an issue pointing at it once this PR is closed.